components/places/src/db/coop_transaction.rs

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::db::PlacesDb;
+use crate::error::*;
+use rusqlite::{Connection, TransactionBehavior};
+use sql_support::{ConnExt, UncheckedTransaction};
+use std::ops::Deref;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+
+/// This implements "cooperative transactions" for places. It relies on our
+/// decision to have exactly 1 general purpose "writer" connection and exactly
+/// one "sync writer" - ie, exactly 2 write connections.
+///
+/// The idea is that anything that uses the sync connection should use
+/// `time_chunked_transaction`. Code using this should regularly call
+/// `maybe_commit()`, and every second, will commit the transaction and start
+/// a new one.
+///
+/// This means that in theory the other writable connection can start
+/// transactions and should block for a max of 1 second - well under the 5
+/// seconds before that other writer will fail with a SQLITE_BUSY or similar error.
+///
+/// However, in practice we see the writer thread being starved - even though
+/// it's waiting for a lock, the sync thread still manages to re-get the lock.
+/// In other words, the locks used by sqlite aren't "fair".
+///
+/// So we mitigate this with a simple approach that works fine within our
+/// "exactly 2 writers" constraints:
+/// * Each database connection shares a mutex.
+/// * Before starting a transaction, each connection locks the mutex.
+/// * It then starts an "exclusive" transaction - because sqlite now holds a
+///   lock on our behalf, we release the lock on the mutex.
+///
+/// In other words, the lock is held only while obtaining the DB lock, then
+/// immediately released.
+///
+/// The end result here is that if either connection is waiting for the
+/// database lock because the other already holds it, the waiting one is
+/// guaranteed to get the database lock next.
+///
+/// One additional wrinkle here is that even if there was exactly one writer,
+/// there's still a possibility of SQLITE_BUSY if the database is being
+/// checkpointed. So we handle that case and perform exactly 1 retry.
```

Comment:
Hmm, I wasn't sure if 1 retry would be sufficient...are you thinking just give up in that case, it's not worth looping?

Reply:
I should have mentioned this - I'm assuming that the 5 second "wait for lock" still applies to both calls even in the checkpoint case. I can't find any docs on that, but if it's true, then we have a 10 second wait already. The logging was supposed to help make that clear (although actually seeing the logs remains an issue).

Sadly I couldn't work out how to simulate that in the example.